### PR TITLE
chore: add middlewares for express, koa, hapi

### PR DESCRIPTION
**Pull request by steplov**
Hello, 
as was discussed in [https://github.com/APIs-guru/graphql-voyager/issues/26](https://github.com/APIs-guru/graphql-voyager/issues/26) i've created the first middleware implementation. 
Let's discuss what is missing and what should be changed.
P.S. there are missing readme and build tasks in root package.json

### CODE REVIEW

**Comment by IvanGoncharov** [comment]
@steplov Wow, you wrote a lot of code 😮 

I think it's better to have one universal middleware for all servers under the main package.
So what about moving `middlewares/common/src/render-voyager-middleware-page.ts` to `src/middleware.ts` and replace `<script src="/voyager/client/voyager.min.js"></script>` to also use `jsdelivr` CDN?
That way you don't need to use `static` middleware and `Router` and end up with a very simple middleware. 

For example middleware for express will look like this:
```js
export function expressMiddleware(options) {
  return (_, res) => {
    res.status(200)
    res.send(renderVoyagerHTML(options))
  }
}
```
As you can see `expressMiddleware` has no extenal dependencies even on `express` package.
So you put 3 functions (express, koa, hapi) together with `renderVoyagerHTML` into `middleware.ts` and distribute it as part of the main package.

The downside is that it wouldn't work without internet access but you already depend on CDN for the other packages.

**Comment by steplov** [comment]
@IvanGoncharov hi, thank you for the answer. 
It was exactly my first idea and implementation to keep them together. I agree it really would be nice to have universal middleware for all servers or to keep it as simple as possible. Unfortunately it didn't work for me (at least in Chrome, i didn't try in other browsers), because Chrome throws next error 
`Uncaught DOMException: Failed to construct 'Worker': Script at 'http://cdn.jsdelivr.net/npm/graphql-voyager@1.0.0-rc.3/dist/voyager.worker.js' cannot be accessed from origin 'http://localhost:3001'.`
I will check the place where code loads worker file, There is an option to change the function that handle it. If it will work - then i will simplify the code.

About external dependencies. We need them mostly for typescript type definitions. How to define types for middleware without them?  

Btw, don't you think that it is bad idea to rely on external cdn in middleware ?

**Comment by RomanGotsiy** [comment]
Hey @steplov,

First, thank you very much for kicking this off!

Regarding the issue with worker loading. There was [an issue](https://github.com/APIs-guru/graphql-voyager/issues/19) opened and I have fixed it already. Version 1.0.0-rc.4 doesn't have it already!

**Comment by steplov** [comment]
@RomanGotsiy , thank you, then the code could be much simpler. I will refactor it and come it a few days

**Comment by IvanGoncharov** [comment]
> could you please deploy 1.0.0-rc.4 to CDN ?

@steplov Done.

> About external dependencies. We need them mostly for typescript type definitions. How to define types for middleware without them?

You can put TS typings as devDependencies of the main package.

> Btw, don't you think that it is bad idea to rely on external cdn in middleware ?

[It's done by](https://github.com/graphql/express-graphql/blob/master/src/renderGraphiQL.js#L66) `express-graphql` to support GraphiQL and it works. 

A possible alternative would make a separate bundle with all dependencies included and then inline JS and CSS directly into HTML returned by middleware. However, I think we should try to keep it simple and think of alternatives only if CDN will cause some real issues.

**Comment by steplov** [comment]
Hello, just finished refactoring. With version 1.0.0-rc.4 it looks much simpler. Could you please check how it works for you?

**Comment by IvanGoncharov** [comment]
@steplov PR looks good and I ❤️  that you add documentation and even examples of graphql servers.
Can you please make separate PR for `.editorconfig`?

**Comment by steplov** [comment]
Done

File: README.md [hunk 1]
```diff
@@ -121,6 +121,86 @@ Build for the web with [webpack](https://webpack.js.org/) ([example](./example/w
 
 **Important:** make sure to copy `voyager.worker.js` from `node_modules/graphql-voyager/dist` to the same folder as your main bundle.
 
+## Middleware
+Graphql Voyager has middleware for next frameworks:
```
Comment (RomanGotsiy) [inline README.md]: minor grammar: for **the**  next frameworks

File: README.md [hunk 2]
```diff
@@ -121,6 +121,86 @@ Build for the web with [webpack](https://webpack.js.org/) ([example](./example/w
...
+
+### Express
+
+#### Properties
+Middleware for Express supports following properties:
```
Comment (RomanGotsiy) [inline README.md]: Minor grammar, change to: `Express middleware supports the following properties:`

File: README.md [hunk 3]
```diff
@@ -121,6 +121,86 @@ Build for the web with [webpack](https://webpack.js.org/) ([example](./example/w
...
+
+app.use('/voyager', middleware({ endpointUrl: '/graphql' }));
+
+app.listen(3000);
+```
+
+### Hapi
+
+#### Properties
+Middleware for Hapi supports following properties:
```
Comment (RomanGotsiy) [inline README.md]: Minor grammar, change to: `Hapi middleware supports the following properties:`

File: README.md [hunk 4]
```diff
@@ -121,6 +121,86 @@ Build for the web with [webpack](https://webpack.js.org/) ([example](./example/w
...
+      endpointUrl: '/graphql',
+    }
+  }
+},() => server.start());
+```
+
+### Koa
+
+#### Properties
+Middleware for Koa supports following properties:
```
Comment (RomanGotsiy) [inline README.md]: Minor grammar, change to: `Koa middleware supports the following properties:`

Comment (steplov) [inline README.md]: I've never been strong in English. Thanks for the review. I'll push the fix when I have the internet connection

File: src/middleware/hapi.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+const pkg = require('../package.json');
+
+interface Register {
+  (server: Server, options: any, next: any): void;
+  attributes?: any;
+}
+
+const hapi: Register = function(server, options, next) {
+  if (!options || !options.voyagerOptions) {
+    throw new Error('Voyager middleware requires options.');
```
Comment (steplov) [inline src/middleware/hapi.ts]: Good point, i will change it

File: src/middleware/express.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import * as express from 'express';
+import renderVoyagerMiddlewarePage, { VoyagerMiddlewareOptions } from './render-voyager-middleware-page';
+
+const { version } = require('../package.json');
```
Comment (steplov) [inline src/middleware/express.ts]: I prefer to keep render-voyager-middleware-page.ts without dependecies. moreover
if you propose to export it from index.js , in my opinion, it makes even more sense not to move dependency inside
> It makes sense to also export renderVoyagerMiddlewarePage function

Comment (IvanGoncharov) [inline src/middleware/express.ts]: @steplov Good point, agree 👍

File: src/middleware/render-voyager-middleware-page.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
+export interface VoyagerMiddlewareOptions {
+    endpointUrl: string;
+    version: string;
+}
+
+export default function renderVoyagerMiddlewarePage(data: VoyagerMiddlewareOptions) {
+  return `
```
Comment (steplov) [inline src/middleware/render-voyager-middleware-page.ts]: So i finished with these changes. Hope this was the last change.

Comment (IvanGoncharov) [inline src/middleware/render-voyager-middleware-page.ts]: @steplov Great 👍 Looking forward to your commit.

> Hope this was the last change.

I reviewed code one more time and don't see any other issues. So after you commit fixes I will merge PR.

**Comment by IvanGoncharov** [comment]
@steplov I merged PR + made some small adjustments, see 96009e8
+ I updated NPM package.

Thank you a lot for this feature 🎆 🎉

**Comment by steplov** [comment]
Cool. Thanks a lot. 👍

**Comment by mkochendorfer** [comment]
It appears the middleware does not work for HAPI 17: https://github.com/APIs-guru/graphql-voyager/issues/65

**Comment by kub1x** [comment]
Hey, sometimes I'm working offline with our instance of graphql server on localhost - thus no CDN. We're using the voyager express middleware. I see in the sources the CDN links are harcoded in the inline index.html code. Is there a way to serve the voyager files directly from the express instance?

**Comment by IvanGoncharov** [comment]
@kub1x Would be cool but I don't think middleware should implicitly serve files 😞 
If you figure out how to do it correctly and reliably PR is welcome.

**Pull request merged**